server/api.go

```diff
@@ -503,7 +552,7 @@ func (a *apiServer) subscribe(ctx context.Context, partition *partition,
 		headersBuf := make([]byte, 28)
 		for {
 			// TODO: this could be more efficient.
-			m, offset, timestamp, _, err := reader.ReadMessage(ctx, headersBuf)
+			m, offset, timestamp, _, err := reader.ReadMessage(ctx, headersBuf, partition.setReadonly)
```

Comment:
Does `partition.setReadonly` need to be protected by a mutex here since the channel is modified concurrently and set within the mutex?

Reply:
That's a good point. I'm not sure of the best way to solve this. Making the partition's internal mutex accessible from the outside doesn't sound great, neither is read locking it during the call to `reader.ReadMessage`.